### ISSUE TITLE
process waits spawn_opt/5 more than 4 hours

### ISSUE THREAD

**Issue by vyacheslavik** [erlang/otp#4989]
A gen_server process is stuck when executing spawn/4 call. Process info shows spawn_opt/5 as a current function and status is waiting:

```
> process_info(P).
[{current_function,{erlang,spawn_opt,5}},
 {status,waiting},
 {message_queue_len,13},
 {trap_exit,false},
 {priority,normal},
 ...]
```

I checked 13 messages of the gen_server, they were all specific to my application - no "{spawn_reply, Ref, ok|error, Pid|Error}", just usual '$gen_cast' and 'system' tagged messages. Judging from messages, the caller was blocked for about 4 hours when I noticed that. I haven't increased net_ticktime.

Current stacktrace looks like:

```
> process_info(P, current_stacktrace).
{current_stacktrace,[{erlang,spawn_opt,5,[]},
                     {erlang,spawn,4,[]},
                     ...
                     {my_gen_server,handle_cast,2,
                                    [{file,"..."},
                                     {line,...}]},
                     {gen_server,try_dispatch,4,
[{file,"gen_server.erl"},{line,695}]},
                     {gen_server,handle_msg,6,
[{file,"gen_server.erl"},{line,771}]},
                     {proc_lib,init_p_do_apply,3,
[{file,"proc_lib.erl"},{line,226}]}]}
```

The code is simple: gen_server just spawns a function on several connected nodes. The function doesn't interact with the gen_server that calls spawn/4. The code worked for a couple of years: I can't be 100% sure if this happened before or not, but in any case I see this situation for the first time. A recent change of my system is upgrade to OTP24.

```
handle_cast(do_start, State) ->
    try
        my_start()
    catch
        What:Why:Stacktrace ->
            logger:warning("~p:~p~n~p", [What, Why, Stacktrace], #{mod => ?MODULE, line => ?LINE})
    end,
    {noreply, State};

my_start() ->
    [spawn(Node, ?MODULE, my_action, [erlang:system_time(1)]) || Node <- [node() | my_cluster_nodes()]].

my_cluster_nodes() ->
    {L, _} = lists:unzip(ets:tab2list(?TBL)),
    L.

my_action(Now) ->
    % work with ETS, use an external database service
```

All nodes of the cluster are ordinary Erlang node, and each node runs the same version of Erlang/OTP 24.0.2:

`Erlang/OTP 24 [erts-12.0.2] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [jit]`

OS is Ubuntu 20 LTS.

The environment is Kubernetes with istio used for networking, that means that the code runs inside docker container, and networking is complicated. It's quite possible that one of nodes of the cluster was restarted abruptly, and may be it was related to version upgrade of istio networking, so we have either restart of a node or a possible glitch of networking to break connection.

**Comment by rickard-green** [comment]
Are you able to reproduce it?

If so, you can use `net_kernel:monitor_nodes()` to log when connections are set up and taken down. Call `net_kernel:monitor_nodes()` in a separate process and log `{nodeup, Node}` and `{nodedown, Node}` messages together with the system time when the message is received to disk.

If the connection is taken down while a process is blocked in `spawn_opt()`, the caller of `spawn_opt()` should receive a `{spawn_reply, Ref, error, noconnection}` message at approximately the same time as your logging process receives a `{nodedown, Node}` for the corresponding node.

If the connection to the corresponding node remains up, while `spawn_opt()` is blocking for a very long time. Check if other communication comes through both ways between the nodes or not. For example, call `erpc:call(Node, erlang, node, [])` from the node with the hanging `spawn_opt()` where `Node` is the same as in the `spawn_opt()` call and verify that it returns `Node`.

**Comment by vyacheslavik** [comment]
So far I was not able to reproduce this issue.

I think it may be explained by the fact that voluntarily I can break connection between nodes only in a predictable and conventional way. My user rights in the environment where I observed this situation are limited: I'm not able to kill docker container neither with erl process (my application) nor with istio (networking).

I tried to reproduce the issue just by killing `erl` process on one of cluster nodes from inside the docker container shell, but this is handled correctly, and I observed only expected behavior.

**Comment by v-lev** [comment]
The issue did reproduce itself for a cluster of 3 nodes that is running the version of code without `net_kernel:monitor_nodes()`, so I'm not able to tell if the connection was taken down or remained up at the moment of `spawn_opt()` call.

The process is blocked in `spawn_opt()` for ~40 hours, the status is `waiting`. There is no `{spawn_reply, Ref, error, noconnection}` message in the queue of the process.

The node of the process is connected to other nodes. There are no signs of connection problems in app logs, and `erpc:call(Node, erlang, node, [])` works properly for 2 sibling nodes and for the node itself -- proper values are returned for all 3 nodes of the cluster when I call `erpc:call` from the node of the blocked process.

All nodes are very actively sharing jobs, CPU is loaded ~50%, inter-node calls are very frequent. I would expect error messages in app logs if there were any connection problems between nodes for more than 30 seconds. However app logs show no errors.

I didn't change anything yet and the cluster is in the same state. Is there anything more that I can check for the live cluster with the blocked process?

**Comment by garazdawi** [comment]
I don't suppose there is any way for you to know if the process was spawned on the other node? Does the process that hangs have a monitor setup towards a process on the other node?

I can't think of any information that you could fetch without using gdb on a core file.

**Comment by vyacheslavik** [comment]
There is no monitor, so we don't know if the process was spawned. With `erlang:process_display(..., backtrace).` I see only that `spawn` was going to spawn the process on a remote node, not locally.

I tried gdb, but get "erts/etc/unix/etp-commands.in:4765: Error in sourced command file: No symbol "UWord" in current context.". I guess that without debug symbols there is not much that we can do remotely.

**Comment by garazdawi** [comment]
There is indeed not much that can be done without debug symbols.

**Comment by rickard-green** [comment]
The bug has been found now. For more information see [comment in issue 5346](https://github.com/erlang/otp/issues/5346#issuecomment-962784021).

**Comment by rickard-green** [comment]
This has been fixed in the just released patch packages [OTP 24.1.5](https://erlang.org/download/OTP-24.1.5.README), [OTP 23.3.4.9](https://erlang.org/download/OTP-23.3.4.9.README), and [OTP 22.3.4.23](https://erlang.org/download/OTP-22.3.4.23.README) now.
